File: AFRi-Studio/design_engine/configurations/schema.py
"""Design configuration schema.

The single source of truth for what a design *is*. Every numeric field carries
UI metadata (range, step, unit, description) so the frontend can build the
parameter editor generatively from ``/api/schema`` -- one definition, no drift
between backend validation and frontend controls.
"""
from __future__ import annotations

import hashlib
import json
from enum import Enum
from typing import Any, Literal

from pydantic import BaseModel, Field, field_validator


# --------------------------------------------------------------------------
# UI metadata helper
# --------------------------------------------------------------------------
def P(default, lo, hi, step, unit, desc, group="general"):
    """A numeric parameter carrying the metadata the UI needs to render it."""
    return Field(
        default,
        ge=lo,
        le=hi,
        json_schema_extra={
            "ui": {
                "min": lo, "max": hi, "step": step,
                "unit": unit, "description": desc, "group": group,
            }
        },
    )


class SplitType(str, Enum):
    BALANCED = "balanced"
    S_RIVER = "s_river"
    ORGANIC = "organic"


class RenderQuality(str, Enum):
    PREVIEW = "preview"
    STANDARD = "standard"
    HIGH = "high"


class CameraPreset(str, Enum):
    TOP = "top"
    FRONT = "front"
    SIDE = "side"
    THREE_QUARTER = "three_quarter"
    CLOSEUP = "closeup"
    HERO = "hero"


class LightingPreset(str, Enum):
    STUDIO_SOFT = "studio_soft"
    DRAMATIC = "dramatic"
    FLAT_CATALOG = "flat_catalog"


class HatStyle(str, Enum):
    FEDORA = "fedora"
    BOATER = "boater"
    WIDE_BRIM = "wide_brim"
    CLOCHE = "cloche"
    BUCKET = "bucket"


class HatMaterial(str, Enum):
    FELT = "felt"
    STRAW = "straw"
    WOOL = "wool"
    CANVAS = "canvas"


class ConformMode(str, Enum):
    RIGID = "rigid"
    SURFACE_CONFORM = "surface_conform"


class AttachmentType(str, Enum):
    NONE = "none"
    PIN = "pin"


class MaterialPreset(str, Enum):
    MATTE_FELT = "matte_felt"
    SATIN_SILK = "satin_silk"
    BRUSHED_METAL = "brushed_metal"
    ENAMEL = "enamel"
    VELVET = "velvet"


# --------------------------------------------------------------------------
# Flower
# --------------------------------------------------------------------------
class FlowerConfig(BaseModel):
    """Parameters of the canonical master marigold.

    Defaults are *provisional studio assumptions*, not client-confirmed values.
    """
    model_config = {"extra": "forbid"}

    diameter_mm: float = P(90.0, 30.0, 200.0, 1.0, "mm",
                           "Overall diameter of the finished flower.", "form")
    relief_depth_mm: float = P(18.0, 4.0, 60.0, 0.5, "mm",
                               "Height of the domed rosette above its base plane.", "form")
    thickness_mm: float = P(1.1, 0.3, 5.0, 0.1, "mm",
                            "Material thickness of each petal.", "form")

    layer_count: int = P(7, 1, 12, 1, "layers",
                         "Number of concentric petal rows.", "petals")
    petal_count_base: int = P(21, 5, 60, 1, "petals",
                              "Petals in the outermost row; inner rows scale from this.", "petals")
    petal_density: float = P(1.25, 0.4, 2.5, 0.05, "x",
                             "Multiplier on petals per row. Higher reads fuller.", "petals")
    petal_length_ratio: float = P(0.46, 0.15, 0.75, 0.01, "x",
                                  "Petal length as a fraction of flower radius.", "petals")
    petal_width_ratio: float = P(1.02, 0.20, 1.40, 0.01, "x",
                                 "Petal width relative to its natural spacing. >1 overlaps.", "petals")
    petal_overlap: float = P(0.34, 0.0, 0.8, 0.01, "x",
                             "Radial overlap between adjacent rows.", "petals")
    petal_curvature: float = P(0.62, 0.0, 1.0, 0.01, "x",
                               "Lengthwise curl. 0 is flat, 1 curls strongly inward.", "shape")
    petal_cup: float = P(0.45, 0.0, 1.0, 0.01, "x",
                         "Crosswise cupping, the channel along the petal.", "shape")
    petal_ruffle_amp: float = P(0.24, 0.0, 1.0, 0.01, "x",
                                "Amplitude of the ruffled petal edge.", "shape")
    petal_ruffle_freq: float = P(3.6, 1.0, 10.0, 0.1, "cycles",
                                 "Number of ruffle waves across the petal edge.", "shape")
    petal_notch: float = P(0.16, 0.0, 0.7, 0.01, "x",
                           "Depth of the notch at the petal tip.", "shape")
    layer_tilt_gain: float = P(0.62, 0.0, 1.4, 0.01, "x",
                               "How much more upright each inner row stands.", "shape")

    center_diameter_ratio: float = P(0.22, 0.08, 0.6, 0.01, "x",
                                     "Centre boss diameter as a fraction of flower diameter.", "center")
    center_dome_height: float = P(0.30, 0.0, 1.2, 0.01, "x",
                                  "Centre dome height relative to relief depth.", "center")
    center_floret_rings: int = P(5, 0, 10, 1, "rings",
                                 "Rings of tiny disc florets in the centre.", "center")

    base_disc_ratio: float = P(0.40, 0.15, 0.95, 0.01, "x",
                               "Structural base disc diameter, fraction of flower diameter. "
                               "This is what physically carries each half.", "structure")
    base_thickness_mm: float = P(1.8, 0.5, 6.0, 0.1, "mm",
                                 "Thickness of the structural base disc.", "structure")

    organic_variation: float = P(0.35, 0.0, 1.0, 0.01, "x",
                                 "Per-petal random jitter in angle, length, width and tilt.", "variation")
    seed: int = Field(20260916, ge=0, le=2**31 - 1, json_schema_extra={
        "ui": {"min": 0, "max": 2**31 - 1, "step": 1, "unit": "",
               "description": "Random seed. Same seed reproduces the flower exactly.",
               "group": "variation"}})

    # Tessellation density -- affects fidelity and cost, not shape.
    petal_segments_u: int = P(18, 5, 40, 1, "segs",
                              "Tessellation along petal length.", "quality")
    petal_segments_v: int = P(11, 3, 25, 1, "segs",
                              "Tessellation across petal width.", "quality")


# --------------------------------------------------------------------------
# Split
# --------------------------------------------------------------------------
class ControlPoint(BaseModel):
    model_config = {"extra": "forbid"}
    y: float = Field(..., ge=-1.5, le=1.5, description="Normalised position along the split axis.")
    x: float = Field(..., ge=-1.5, le=1.5, description="Normalised lateral offset.")


class SplitConfig(BaseModel):
    """How the master flower is divided into two pieces."""
    model_config = {"extra": "forbid"}

    type: SplitType = Field(SplitType.S_RIVER, json_schema_extra={
        "ui": {"description": "Family of dividing curve.", "group": "split"}})
    position: float = P(0.0, -0.6, 0.6, 0.01, "x",
                        "Lateral position of the split. 0 is centred.", "split")
    orientation_deg: float = P(0.0, -180.0, 180.0, 1.0, "deg",
                               "Rotation of the split axis around the flower.", "split")
    amplitude: float = P(0.30, 0.0, 0.9, 0.01, "x",
                         "How far the curve swings sideways. Drives the S.", "curve")
    smoothness: float = P(0.65, 0.05, 1.0, 0.01, "x",
                          "Curve softness. Low is taut and angular, high is flowing.", "curve")
    control_points: list[ControlPoint] | None = Field(None, json_schema_extra={
        "ui": {"description": "Explicit Bezier control points. Overrides amplitude/smoothness "
                              "when present, for hand-authored curves.", "group": "curve"}})

    organic_octaves: int = P(3, 1, 6, 1, "octaves",
                             "Noise detail levels for the organic split.", "organic")
    organic_roughness: float = P(0.45, 0.0, 1.0, 0.01, "x",
                                 "Irregularity of the organic split.", "organic")
    organic_seed: int = Field(7, ge=0, le=2**31 - 1, json_schema_extra={
        "ui": {"min": 0, "max": 2**31 - 1, "step": 1, "unit": "",
               "description": "Seed for the organic split's wander.", "group": "organic"}})

    separation_mm: float = P(14.0, 0.0, 80.0, 0.5, "mm",
                             "Gap between the pieces in the separated view.", "presentation")
    boundary_tolerance_mm: float = P(0.05, 0.001, 1.0, 0.001, "mm",
                                     "Tolerance for boundary and contamination checks.", "validation")
    cap_boundary: bool = Field(True, json_schema_extra={
        "ui": {"description": "Build a real boundary wall on the cut face so each "
                              "piece is a closed solid.", "group": "split"}})

    @field_validator("control_points")
    @classmethod
    def _cp_monotone(cls, v):
        if v is None:
            return v
        if len(v) < 2:
            raise ValueError("need at least 2 control points")
        if len(v) > 8:
            raise ValueError("at most 8 control points")
        ys = [p.y for p in v]
        if any(b <= a for a, b in zip(ys, ys[1:])):
            raise ValueError(
                "control point y values must strictly increase: the split path must "
                "remain a function of y so that it divides the plane into exactly two regions"
            )
        return v



# --------------------------------------------------------------------------
# Hat (stage two)
# --------------------------------------------------------------------------
class HatConfig(BaseModel):
    """The hat the accessory is worn on.

    Head circumference is the one dimension a hat cannot get wrong, so the
    geometry is derived from it rather than from a guessed radius. Defaults are
    a standard adult 58 cm fedora; like everything else here they are studio
    assumptions, not AFRi specifications.
    """
    model_config = {"extra": "forbid"}

    style: HatStyle = Field(HatStyle.FEDORA, json_schema_extra={
        "ui": {"description": "Hat silhouette. Sets crown taper, dome, flare and edge softness.",
               "group": "hat"}})
    head_circumference_mm: float = P(580.0, 520.0, 640.0, 5.0, "mm",
                                     "Inside head circumference. 58 cm is the commonest adult size.", "hat")
    crown_height_mm: float = P(112.0, 40.0, 200.0, 1.0, "mm",
                               "Height of the crown above the brim plane.", "hat")
    crown_taper: float = P(0.0, 0.0, 1.1, 0.01, "x",
                           "Crown top radius as a fraction of head radius. 0 uses the style default.", "hat")
    crown_crease: float = P(1.0, 0.0, 1.5, 0.01, "x",
                            "Depth of the centre crease and finger dents. 0 leaves a smooth crown.", "hat")

    brim_width_mm: float = P(68.0, 10.0, 180.0, 1.0, "mm",
                             "Brim reach beyond the head radius.", "brim")
    brim_droop_deg: float = P(9.0, -10.0, 45.0, 0.5, "deg",
                              "How far the brim falls from horizontal.", "brim")
    brim_curl: float = P(0.10, 0.0, 0.6, 0.01, "x",
                         "Upward curl at the outer edge.", "brim")
    thickness_mm: float = P(1.6, 0.4, 6.0, 0.1, "mm",
                            "Material thickness of the hat body.", "brim")

    band_depth_mm: float = P(2.2, 0.0, 12.0, 0.1, "mm",
                             "How far the ribbon band stands off the crown. 0 omits the band.", "band")
    band_height_mm: float = P(34.0, 0.0, 90.0, 1.0, "mm",
                              "Height of the ribbon band.", "band")
    band_z_mm: float = P(3.0, 0.0, 60.0, 0.5, "mm",
                         "Height of the band's lower edge above the brim plane.", "band")

    material: HatMaterial = Field(HatMaterial.FELT, json_schema_extra={
        "ui": {"description": "Hat material. Drives roughness and sheen at render time.",
               "group": "hat"}})
    base_color: str = Field("#2C2A27", pattern=r"^#[0-9a-fA-F]{6}$",
                            description="Provisional charcoal felt. NOT an AFRi brand colour.")
    band_color: str = Field("#171513", pattern=r"^#[0-9a-fA-F]{6}$",
                            description="Ribbon band colour.")

    profile_segments: int = P(200, 60, 400, 10, "segs",
                              "Samples along the meridian. Affects fidelity and cost, not shape.", "quality")
    revolve_segments: int = P(128, 32, 320, 8, "segs",
                              "Samples around the axis.", "quality")


class PlacementConfig(BaseModel):
    """Where the flower sits on the hat, and how it is held there."""
    model_config = {"extra": "forbid"}

    show_hat: bool = Field(True, json_schema_extra={
        "ui": {"description": "Include the hat. Off renders the accessory alone, as in stage one.",
               "group": "placement"}})
    azimuth_deg: float = P(-52.0, -180.0, 180.0, 1.0, "deg",
                           "Position around the crown. 0 is the front of the hat.", "placement")
    radial_position: float = P(0.10, 0.0, 1.0, 0.01, "x",
                               "0 seats the flower against the band at the crown foot; "
                               "1 puts it at the brim edge.", "placement")
    surface_offset_mm: float = P(0.0, -5.0, 30.0, 0.1, "mm",
                                 "Lift off the hat surface. Negative presses into it.", "placement")
    tilt_deg: float = P(0.0, -60.0, 60.0, 1.0, "deg",
                        "Tip the flower outward (+) or inward (-) from the local surface.", "placement")
    roll_deg: float = P(0.0, -180.0, 180.0, 1.0, "deg",
                        "Spin the flower in its own plane. Rotates where the split line points.", "placement")
    conform: ConformMode = Field(ConformMode.RIGID, json_schema_extra={
        "ui": {"description": "Rigid keeps the flower flat and takes up the gap in the fixing. "
                              "Surface conform is not implemented; see the contact report.",
               "group": "placement"}})

    attachment: AttachmentType = Field(AttachmentType.PIN, json_schema_extra={
        "ui": {"description": "Fixing between each piece and the hat.", "group": "attachment"}})
    pin_count: int = P(2, 1, 4, 1, "pins",
                       "Pins per piece. Two stop a piece rotating about a single pin.", "attachment")
    pin_diameter_mm: float = P(1.6, 0.6, 5.0, 0.1, "mm",
                               "Pin shank diameter.", "attachment")
    pin_length_mm: float = P(9.0, 3.0, 25.0, 0.5, "mm",
                             "Pin length below the base disc. Must exceed the hat thickness.", "attachment")

    contact_tolerance_mm: float = P(0.6, 0.05, 5.0, 0.05, "mm",
                                    "Gap below which the base disc counts as touching the hat.", "validation")


# --------------------------------------------------------------------------
# Material / Render
# --------------------------------------------------------------------------
class MaterialConfig(BaseModel):
    model_config = {"extra": "forbid"}
    preset: MaterialPreset = Field(MaterialPreset.SATIN_SILK)
    base_color: str = Field("#F2A007", pattern=r"^#[0-9a-fA-F]{6}$",
                            description="Provisional marigold orange. NOT an AFRi brand colour.")
    secondary_color: str = Field("#C9500A", pattern=r"^#[0-9a-fA-F]{6}$",
                                 description="Deeper tone used at petal roots and centre.")
    roughness: float = P(0.42, 0.0, 1.0, 0.01, "x", "Surface roughness.", "material")
    metallic: float = P(0.0, 0.0, 1.0, 0.01, "x", "Metallic amount.", "material")
    sheen: float = P(0.35, 0.0, 1.0, 0.01, "x", "Fabric sheen at grazing angles.", "material")
    piece_tint: float = P(0.0, 0.0, 1.0, 0.01, "x",
                          "Tints piece B away from piece A so the division reads clearly "
                          "in presentation images. 0 keeps both identical.", "material")


class RenderConfig(BaseModel):
    model_config = {"extra": "forbid"}
    quality: RenderQuality = Field(RenderQuality.PREVIEW)
    camera: CameraPreset = Field(CameraPreset.THREE_QUARTER)
    lighting: LightingPreset = Field(LightingPreset.STUDIO_SOFT)
    resolution: int = P(900, 256, 2400, 32, "px", "Square output resolution.", "render")
    background: str = Field("#14161A", pattern=r"^#[0-9a-fA-F]{6}$",
                            description="Studio backdrop colour.")
    separated: bool = Field(False, description="Render the pieces apart rather than assembled.")


# --------------------------------------------------------------------------
# Root
# --------------------------------------------------------------------------
ENGINE_VERSION = "1.0.0"


class DesignConfig(BaseModel):
    """The unit of versioning, diffing, hashing and AI mutation."""
    model_config = {"extra": "forbid"}

    flower: FlowerConfig = Field(default_factory=FlowerConfig)
    split: SplitConfig = Field(default_factory=SplitConfig)
    hat: HatConfig = Field(default_factory=HatConfig)
    placement: PlacementConfig = Field(default_factory=PlacementConfig)
    material: MaterialConfig = Field(default_factory=MaterialConfig)
    render: RenderConfig = Field(default_factory=RenderConfig)

    # ---- stage hashing: drives the incremental-regeneration cache -------
    def _hash(self, parts: list[dict]) -> str:
        blob = json.dumps(parts, sort_keys=True, separators=(",", ":"))
        return hashlib.sha256((ENGINE_VERSION + blob).encode()).hexdigest()[:16]

    def hash_flower(self) -> str:
        return self._hash([self.flower.model_dump(mode="json")])

    def hash_split(self) -> str:
        return self._hash([self.flower.model_dump(mode="json"),
                           self.split.model_dump(mode="json")])

    def hash_hat(self) -> str:
        """The hat depends on nothing else, so it caches independently of the
        flower -- restyling the flower must not rebuild the hat."""
        return self._hash([self.hat.model_dump(mode="json")])

    def hash_assembly(self) -> str:
        return self._hash([self.flower.model_dump(mode="json"),
                           self.split.model_dump(mode="json"),
                           self.hat.model_dump(mode="json"),
                           self.placement.model_dump(mode="json")])

    def hash_scene(self) -> str:
        return self._hash([self.flower.model_dump(mode="json"),
                           self.split.model_dump(mode="json"),
                           self.hat.model_dump(mode="json"),
                           self.placement.model_dump(mode="json"),
                           self.material.model_dump(mode="json")])

    def hash_render(self) -> str:
        return self._hash([self.flower.model_dump(mode="json"),
                           self.split.model_dump(mode="json"),
                           self.hat.model_dump(mode="json"),
                           self.placement.model_dump(mode="json"),
                           self.material.model_dump(mode="json"),
                           self.render.model_dump(mode="json")])

    def diff(self, other: "DesignConfig") -> dict[str, dict[str, Any]]:
        """Flat parameter-level diff, for the version comparison view."""
        a, b = self.model_dump(mode="json"), other.model_dump(mode="json")
        out: dict[str, dict[str, Any]] = {}
        for section in a:
            for key in a[section]:
                if a[section][key] != b[section][key]:
                    out[f"{section}.{key}"] = {"from": a[section][key], "to": b[section][key]}
        return out


# --------------------------------------------------------------------------
# UI schema export
# --------------------------------------------------------------------------
def ui_schema() -> dict:
    """Flatten the model metadata into what the parameter editor needs."""
    out: dict[str, Any] = {"engine_version": ENGINE_VERSION, "sections": {}}
    for name, model in (("flower", FlowerConfig), ("split", SplitConfig),
                        ("hat", HatConfig), ("placement", PlacementConfig),
                        ("material", MaterialConfig), ("render", RenderConfig)):
        js = model.model_json_schema()
        fields: list[dict] = []
        for fname, finfo in model.model_fields.items():
            prop = js.get("properties", {}).get(fname, {})
            extra = (finfo.json_schema_extra or {}).get("ui", {}) if isinstance(
                finfo.json_schema_extra, dict) else {}
            ann = finfo.annotation
            kind = "number"
            choices = None
            if isinstance(ann, type) and issubclass(ann, Enum):
                kind, choices = "enum", [e.value for e in ann]
            elif ann is bool:
                kind = "bool"
            elif ann is int:
                kind = "int"
            elif ann is str:
                kind = "color" if "pattern" in prop and "0-9a-fA-F" in prop.get("pattern", "") else "text"
            elif fname == "control_points":
                kind = "curve_points"
            default = finfo.default
            if isinstance(default, Enum):
                default = default.value
            if default is not None and not isinstance(default, (int, float, str, bool, list)):
                default = None
            fields.append({
                "name": fname, "kind": kind, "default": default, "choices": choices,
                "min": extra.get("min"), "max": extra.get("max"), "step": extra.get("step"),
                "unit": extra.get("unit", ""),
                "description": extra.get("description") or prop.get("description", ""),
                "group": extra.get("group", "general"),
            })
        out["sections"][name] = fields
    return out

File: AFRi-Studio/web_studio/schema.js
/* AFRi Studio -- parameter schema, concepts, finishes and framing.
 *
 * Everything the interface can change about a design is declared here, once,
 * with its bounds and the sentence that explains it. The controls, the command
 * palette, the Claude designer's allow-list and the tech pack all read this
 * table, so a parameter cannot exist in one and be missing from another.
 */

/* group = [title, openByDefault, [ [key, min, max, step, unit, help], ... ] ] */
export const FLOWER_GROUPS = [
  ['Form', true, [
    ['diameter_mm', 30, 200, 1, 'mm', 'Overall diameter of the finished flower.'],
    ['relief_depth_mm', 4, 60, 0.5, 'mm', 'Height of the domed rosette above its base plane.'],
    ['thickness_mm', 0.3, 5, 0.1, 'mm', 'Material thickness of each petal.'],
  ]],
  ['Petals', true, [
    ['layer_count', 1, 12, 1, 'rows', 'Number of concentric petal rows.'],
    ['petal_count_base', 5, 60, 1, 'petals', 'Petals in the outermost row; inner rows scale from this.'],
    ['petal_density', 0.4, 2.5, 0.05, '×', 'Multiplier on petals per row. Higher reads fuller.'],
    ['petal_length_ratio', 0.15, 0.75, 0.01, '×', 'Petal length as a fraction of flower radius.'],
    ['petal_width_ratio', 0.2, 1.4, 0.01, '×', 'Width relative to natural spacing. Above 1 overlaps.'],
    ['petal_overlap', 0, 0.8, 0.01, '×', 'Radial overlap between adjacent rows.'],
  ]],
  ['Petal shape', false, [
    ['petal_curvature', 0, 1, 0.01, '×', 'Lengthwise curl. 0 is flat, 1 curls strongly inward.'],
    ['dome_gain', 0.1, 1.2, 0.01, '×', 'How steeply the rows climb toward the centre. This is what turns a flat rosette into a pompon.'],
    ['petal_cup', 0, 1, 0.01, '×', 'Crosswise cupping, the channel along the petal.'],
    ['petal_ruffle_amp', 0, 1, 0.01, '×', 'Amplitude of the ruffled petal edge.'],
    ['petal_ruffle_freq', 1, 10, 0.1, 'cycles', 'Number of ruffle waves across the petal edge.'],
    ['petal_notch', 0, 0.7, 0.01, '×', 'Depth of the notch at the petal tip.'],
    ['layer_tilt_gain', 0, 1.4, 0.01, '×', 'How much more upright each inner row stands.'],
  ]],
  ['Centre', false, [
    ['center_diameter_ratio', 0.08, 0.6, 0.01, '×', 'Centre boss diameter as a fraction of flower diameter.'],
    ['center_dome_height', 0, 1.2, 0.01, '×', 'Centre dome height relative to relief depth.'],
    ['center_floret_rings', 0, 10, 1, 'rings', 'Rings of tiny disc florets in the centre.'],
  ]],
  ['Structure', false, [
    ['base_disc_ratio', 0.15, 0.95, 0.01, '×', 'Structural base disc diameter. This is what physically carries each half.'],
    ['base_thickness_mm', 0.5, 6, 0.1, 'mm', 'Thickness of the structural base disc.'],
  ]],
  ['Variation', false, [
    ['organic_variation', 0, 1, 0.01, '×', 'Per-petal jitter in angle, length, width and tilt.'],
    ['seed', 1, 999999, 1, '', 'Random seed. The same seed reproduces the flower exactly.'],
  ]],
];

export const SPLIT_GROUP = ['Dividing curve', true, [
  ['position', -0.6, 0.6, 0.01, '×', 'Lateral position of the split. 0 is centred.'],
  ['orientation_deg', -180, 180, 1, '°', 'Rotation of the split axis around the flower.'],
  ['amplitude', 0, 0.9, 0.01, '×', 'How far the curve swings sideways. This drives the S.'],
  ['smoothness', 0.05, 1, 0.01, '×', 'Curve softness. Low is taut and angular, high is flowing.'],
  ['organic_roughness', 0, 1, 0.01, '×', 'Irregularity of the organic split only.'],
]];

export const HAT_GROUP = ['Hat', false, [
  ['head_circumference_mm', 520, 640, 5, 'mm', 'Inside head circumference. 58 cm is the commonest adult size, and it is the one dimension a hat cannot get wrong.'],
  ['crown_height_mm', 40, 200, 1, 'mm', 'Height of the crown above the brim plane.'],
  ['crown_crease', 0, 1.5, 0.01, '×', 'Depth of the centre crease and finger dents. 0 leaves a smooth crown.'],
  ['brim_width_mm', 10, 180, 1, 'mm', 'Brim reach beyond the head radius. This is what decides whether the accessory fits.'],
  ['brim_droop_deg', -10, 45, 0.5, '°', 'How far the brim falls from horizontal.'],
  ['brim_curl', 0, 0.6, 0.01, '×', 'Upward curl at the outer edge.'],
  ['thickness_mm', 0.4, 6, 0.1, 'mm', 'Material thickness of the hat body.'],
  ['band_height_mm', 0, 90, 1, 'mm', 'Height of the ribbon band. 0 omits it.'],
]];

export const PLACEMENT_GROUP = ['Placement', false, [
  ['radial_position', 0, 1, 0.01, '×', '0 seats the accessory against the band at the crown foot; 1 puts it at the brim edge.'],
  ['azimuth_deg', -180, 180, 1, '°', 'Position around the crown. 0 is the front of the hat.'],
  ['surface_offset_mm', -5, 30, 0.1, 'mm', 'Lift off the hat surface. The fixing takes up this gap.'],
  ['tilt_deg', -60, 60, 1, '°', 'Tip the accessory outward (+) or inward (-) from the local surface.'],
  ['roll_deg', -180, 180, 1, '°', 'Spin it in its own plane. This aims the dividing curve.'],
]];

export const HAT_STYLES = [
  ['fedora', 'Fedora'], ['wide_brim', 'Wide brim'], ['boater', 'Boater'],
  ['cloche', 'Cloche'], ['bucket', 'Bucket'],
];

export const CURVE_FAMILIES = [
  ['balanced', 'Balanced — single stroke'],
  ['s_river', 'S river — meander'],
  ['organic', 'Organic — fractal wander'],
];

export const CONCEPTS = [
  { key: 'A', name: 'Balanced Split',
    desc: 'A clear, near-symmetric division. The dividing line reads as a single confident stroke.',
    split: { type: 'balanced', position: 0, amplitude: 0.18, smoothness: 0.7, orientation_deg: 0, separation_mm: 18 } },
  { key: 'B', name: 'S River Split', priority: true,
    desc: 'A flowing, river-inspired separation with an editable curve.',
    split: { type: 's_river', position: 0, amplitude: 0.34, smoothness: 0.72, orientation_deg: 12, separation_mm: 18 } },
  { key: 'C', name: 'Organic Asymmetric',
    desc: 'An expressive, irregular division that wanders with the petal arrangement.',
    split: { type: 'organic', position: 0.10, amplitude: 0.40, smoothness: 0.55, orientation_deg: -28,
             separation_mm: 55, organic_octaves: 3, organic_roughness: 0.55, organic_seed: 11 } },
];

/* The refined marigold from the Concept C v2 review. One master flower serves
 * all three concepts -- only the dividing curve differs -- so the refinement
 * applies to every concept. ENGINE.FLOWER_DEFAULTS is left alone: it is the
 * engine's own baseline and what the Python parity test compares against. */
export const REFINED = {
  layer_count: 9, petal_count_base: 26, petal_density: 1.65, petal_overlap: 0.56,
  petal_length_ratio: 0.37, petal_width_ratio: 1.12,
  petal_curvature: 0.64, petal_cup: 0.58, layer_tilt_gain: 0.80,
  relief_depth_mm: 40.0, dome_gain: 0.56,
  petal_ruffle_amp: 0.42, petal_ruffle_freq: 4.6, petal_notch: 0.10,
  center_diameter_ratio: 0.14, center_dome_height: 0.16, center_floret_rings: 3,
  organic_variation: 0.52,
};

/* Finishes. Each is a plausible way the accessory could actually be made, and
 * each sets the shading of both halves plus the freshly cut wall between them,
 * which in a real part is the one face that shows the raw material. */
export const MATERIALS = [
  { key: 'resin', name: 'Marigold resin', note: 'Cast urethane, pigment through the body.',
    a: 0xF2A007, b: 0xE0780D, wall: 0xC8541B, roughness: 0.62, metalness: 0.02, wallRough: 0.48 },
  { key: 'brass', name: 'Polished brass', note: 'Lost-wax cast, hand polished.',
    a: 0xC9942F, b: 0xB07B22, wall: 0x8C5F18, roughness: 0.26, metalness: 0.92, wallRough: 0.42 },
  { key: 'enamel', name: 'Cream enamel', note: 'Vitreous enamel over a cast core.',
    a: 0xF3E8D3, b: 0xE4D4B6, wall: 0xB98B4E, roughness: 0.22, metalness: 0.04, wallRough: 0.5 },
  { key: 'felt', name: 'Dyed wool felt', note: 'Blocked felt, the millinery route.',
    a: 0xC4541F, b: 0xA6421A, wall: 0x7C3214, roughness: 0.96, metalness: 0.0, wallRough: 0.96 },
  { key: 'graphite', name: 'Matte graphite', note: 'SLS nylon, dyed and tumbled.',
    a: 0x4A4A46, b: 0x3A3A36, wall: 0x6E6A5E, roughness: 0.82, metalness: 0.12, wallRough: 0.7 },
];

/* Lighting rigs. Colour and intensity only -- the geometry never moves, so a
 * measurement taken under one rig is the measurement under all of them. */
export const ENVIRONMENTS = [
  { key: 'studio', name: 'Studio warm',
    hemi: [0xF6E2C0, 0x1A1610, 0.85],
    key_: [0xFFF0D6, 2.35, [7, -9, 12]],
    fill: [0xC8D8FF, 0.55, [-9, 4, 5]],
    rim: [0xFFB347, 0.85, [-3, 10, -6]] },
  { key: 'daylight', name: 'North daylight',
    hemi: [0xDDE8F6, 0x201E18, 1.15],
    key_: [0xFFFFFF, 1.9, [5, -8, 13]],
    fill: [0xBFD4F2, 1.0, [-8, 5, 6]],
    rim: [0xFFFFFF, 0.45, [-2, 9, -5]] },
  { key: 'dusk', name: 'Dusk',
    hemi: [0x6B5B7A, 0x14110C, 0.5],
    key_: [0xFFC98A, 1.5, [9, -6, 6]],
    fill: [0x7D8BC4, 0.4, [-7, 6, 4]],
    rim: [0xFF8A3D, 1.6, [-4, 8, -7]] },
  { key: 'flat', name: 'Flat inspection',
    hemi: [0xFFFFFF, 1.5, [0xBFC3B4, 0, 0]],
    key_: [0xFFFFFF, 1.1, [6, -7, 10]],
    fill: [0xFFFFFF, 0.9, [-7, 5, 6]],
    rim: [0xFFFFFF, 0.5, [0, 8, -8]] },
];

/* azimuth, elevation, and distance as a multiple of the measured radius, so
 * the framing holds when the flower's diameter changes. */
export const VIEWS = {
  three_quarter: [0.66, 0.52, 4.7],
  top: [-Math.PI / 2, 1.44, 4.3],
  front: [-Math.PI / 2, 0.10, 5.0],
  side: [0, 0.14, 5.0],
};
export const VIEW_LABELS = [
  ['three_quarter', 'Three quarter'], ['top', 'Top'], ['front', 'Front'], ['side', 'Side'],
];

/* ---- flat index -------------------------------------------------------- */
/* name -> {scope, key, lo, hi, step, unit, help, label}. One lookup table for
 * the command palette, the designer's validator and the parameter diff. */
export const PARAMS = (() => {
  const out = new Map();
  const add = (scope, group) => {
    for (const [key, lo, hi, step, unit, help] of group) {
      out.set(scope + '.' + key, { scope, key, lo, hi, step, unit, help, label: label(key) });
    }
  };
  for (const [, , items] of FLOWER_GROUPS) add('flower', items);
  add('split', SPLIT_GROUP[2]);
  add('hat', HAT_GROUP[2]);
  add('placement', PLACEMENT_GROUP[2]);
  return out;
})();

export function label(key) {
  return key.replace(/_/g, ' ').replace(/\bmm\b/, '(mm)').replace(/\bdeg\b/, '(deg)');
}

export function fmt(x, d) { return Number(x).toFixed(d); }

/* A parameter's display value: integers stay integers, everything else keeps
 * two decimals, and the unit rides along. */
export function showValue(spec, v) {
  const s = (spec.step >= 1) ? String(Math.round(v)) : fmt(v, 2);
  return spec.unit ? s + ' ' + spec.unit : s;
}
